Add sniper_rifles mod

ClobberXD opened the pull request:
This mod depends on shooter, but allows the player to use RMB to "scope in" while wielding the guns.

This PR is Ready for review.

Thomas--S commented on mods/pvp/sniper_rifles/README.md:
```diff
@@ -0,0 +1,15 @@
+
+### Attributions
+
+- `sniper_rifles_rifle_762.ogg` (`Public Domain`)
```
[Public Domain Is Not Open Source](https://opensource.org/node/878)

rubenwardy replied:
if you're in a country which recognises public domain, you can relicense it for other peopl

ClobberXD replied:
Aww man...

> if you're in a country which recognises public domain, you can relicense it for other peopl

Too much work. I'll just find a worse-sounding... well, sound.

Panquesito7 commented on mods/pvp/sniper_rifles/mod.conf:
```diff
@@ -0,0 +1,2 @@
+name = sniper_rifles
+depends = shooter
```
Add a description (after this line). 
I think it will be useful.

ClobberXD replied:
Will do

Panquesito7 commented on mods/pvp/sniper_rifles/README.md:
```diff
@@ -0,0 +1,15 @@
+# sniper_rifles
```
```suggestion
# Sniper Rifles
```

Panquesito7 commented on mods/pvp/sniper_rifles/README.md:
```diff
@@ -0,0 +1,15 @@
+
+## Licenses
+
+- Code license: MIT
```
```suggestion
- License for code: MIT
```

ClobberXD replied:
Does this really matter?

Panquesito7 replied:
IMHO, I think it is.
But it also looks good like `Code license`.

ClobberXD replied:
I'd prefer to keep it as it is, as the suggested way is a tad more verbose.

Panquesito7 replied:
Okay. 
Please resolve this conversation.

Panquesito7 commented on mods/pvp/sniper_rifles/README.md:
```diff
@@ -0,0 +1,15 @@
+## Licenses
+
+- Code license: MIT
+- Media license: CC0 1.0
```
```suggestion
- License for media: CC0 1.0
```

Panquesito7 commented on mods/pvp/sniper_rifles/README.md:
```diff
@@ -0,0 +1,15 @@
+### Attributions
+
+- `sniper_rifles_rifle_762.ogg` (`CC0 1.0`)
+  - Converted from [Battle Rifle.wav](https://freesound.org/people/morganpurkis/sounds/391725/) by [morganpurkis](https://freesound.org/people/morganpurkishttps://freesound.org/people/morganpurkis/sounds/391725/)
```
Add a dot at the end of the line.
```suggestion
  - Converted from [Battle Rifle.wav](https://freesound.org/people/morganpurkis/sounds/391725/) by [morganpurkis](https://freesound.org/people/morganpurkishttps://freesound.org/people/morganpurkis/sounds/391725/).
```

ClobberXD replied:
Not strictly required, but sure, thanks.

ClobberXD commented:
Removed the controversially-licensed sound, and added `description` to `mod.conf`. I just need to create another inventory image for the second rifle and the PR would be complete.

MinetestSam commented:
Tested in singleplayer and everything works fine. I would just like to suggest that probably these sniper rifles should be able to destroy reinforced cobble but again I think it could be a bit overpowered.

ClobberXD commented:
This PR now only adds one sniper rifle, although the registration code for the second rifle has just been commented out. This is because I suck at texturing and my wiggly-squishy brain can only cope with one pixelated texture a year.

I'll "re-add" the second rifle, once I make the second texture and find an appropriate sound for it.

ClobberXD commented:
Replaced the placeholder texture with a proper(ish..?) texture. This PR is ready for review.

MinetestSam commented:
Everything works. The textures also look great!

You should totally add magnum back with the similar texture to the current sniper rifle.

ClobberXD commented:
> You should totally add magnum back with the similar texture to the current sniper rifle.

I'd love to, but with the same texture?

MinetestSam commented:
Since the magnum is a bit better it should have something more to the texture to signify that it is better. Probably make the scope a bit golden along with the back of the gun?

ClobberXD commented:
I like the idea of adding highlights to the same texture for the purpose of differentiation - saves me work :D

ClobberXD commented:
I found time to work on the second rifle texture, and even got to finish it. So I've re-added the magnum rifle. PR needs minimalistic testing to verify that it works.

P.S. I've also modified the existing texture a bit, to be able to differentiate the rifles better.

JostP commented:
Tested in singleplayer (multiplayer didn't work for CTF, no idea why, MTG works normal) and everything looks good. I love new rifles :D and their sound, the only thing I don't like is that both rifles have the same sound and they look too similar.

I'm looking forward to see this added to CTF :D

ClobberXD commented:
Yes, both rifles use the exact same sound. The reason is that I couldn't find any free-to-use sound that sounds good. The very few I did find had too much echo or their authors made bad licensing choices, making them unusable in CTF. For an example of this, check out the initial revision of this PR, from which I had removed the second sound due to the sound being in public domain and having no license.

ClobberXD commented:
Rebased. Please review.

Thomas--S commented:
I did a quick review of this.

First of all, the sounds of the new rifles are much too loud. Please make them as loud as the already existing sounds.

Secondly, I'm not quite sure if we really need new guns, or whether it would be sufficient to add the zoom function to the already exisiting rifle.
If you think new ones should be added, I think we should only add a single one to reduce complexity for the players. The textures also need improvement, they don't fit the style of the other ones.

Thirdly, I feel like the globalstep interval for the zoom check is a little bit short. It could probably be increased a bit to improve performance, but I haven't tested this yet.

ClobberXD commented:
> First of all, the sounds of the new rifles are much too loud. Please make them as loud as the already existing sounds

They're supposed to be that way - they're sniper rifles, not squirt guns.

> Secondly, I'm not quite sure if we really need new guns, or whether it would be sufficient to add the zoom function to the already exisiting rifle.

I see your point. But the thing is that guns can't be directly compared to other melee weapons, like swords. Unlike swords, which can only vary in damage, guns have a multitude of properties, all of which can be tuned to create a unique character for each gun. Players like more of everything, and I really doubt something as fundamental as more weapons can cause confusion. And if you still think that 3 rifles are too much, heck, it's the plain ol' rifle that ought to be removed. No really, it's the most non-unique gun in the lot - both damage and rate of fire is the same as the pistol, and even node-destroying capabilities isn't unique to it. I'd prefer to do that in a different PR though, for the record.

Thomas--S commented:
> They're supposed to be that way - they're sniper rifles, not squirt guns.

It'd be fine for me if those are slightly louder than normal guns, but the current situation is too extreme. Players probably set their sound so that a normal gun sound is rather loud and then those new sounds could cause serious harm. Health has to be more important here than realism. (And I guess that real users of such weapons probably wear ear protectors anyway.)

---

I can understand that experienced players want to have more choices. However, I think that for new players having too much options might be rather confusing.
I first tried this PR in game before looking at the code, and I didn't notice which of the two was better/what was the difference. I had to look in the code to find out that one is much better than the other.
As a compromise, I'd suggest to keep the old rifle (without zoom) and add a second one (probably an average of the two from this PR) as a better version with zoom. The texture of this second (new) one could be based on the old rifle, probably with the addition of a telescope.

This way, the difference would be obvious, the texture fits the style of the other ones and there would be more options. I guess that this could be an acceptable compromise.

ClobberXD commented:
> Players probably set their sound so that a normal gun sound is rather loud and then those new sounds could cause serious harm.

I think I understand what you mean now. My volume levels are *very* low (it'd be muted most of the time, though), so I didn't feel that these were loud. I'll see what I can do.

About the textures being out of style, we could make them smaller to fit, similar to how we've made pistols larger. But it's not the size that's the problem - it's the length-to-height ratio of sniper rifles, that make them hard to fit. If we try to increase the size, the muzzle and butt would extend out of the inv. slot's viewport and get clipped. Trying to increase only the height of the rifle would result in an equally-ugly squished texture.

> I first tried this PR in game before looking at the code, and I didn't notice which of the two was better/what was the difference. I had to look in the code to find out that one is much better than the other.

This is a very valid concern. I had originally assumed that the differences in the textures would suffice, but thinking again, I suppose those don't have any significance to players who don't know much about guns. I'm out of ideas, but I'll see what I can do about it.

> As a compromise, I'd suggest to keep the old rifle (without zoom) and add a second one (probably an average of the two from this PR) as a better version with zoom. The texture of this second (new) one could be based on the old rifle, probably with the addition of a telescope.

- I already hate the existing rifle. Please don't make me do this. :P
- A rifle ain't a sniper rifle without its scope.

Maybe we can make do with only one sniper rifle for now?

Thomas--S commented:
> I already hate the existing rifle

Could you please specify what you hate here? The texture? The behaviour?
IMHO, I think the texture at least is rather good. I am also opposed to too realistic textures, so in general I quite like the texture style of the shooter mod.

I am no weapon expert, but what do you think about having one normal rifle similiar (or identical) to the current one and one sniper rifle with zoom (with only the texture being based on the exisiting rifle + a scope)

ClobberXD commented:
> Could you please specify what you hate here? The texture? The behaviour?

All of it :D

The texture looks like a play-toy (yes, MT isn't particularly realistic but this is overkill); both the damage and fire-rate of the rifle is very similar to the pistol; not to mention, both are extremely unrealistic too. Going by the texture, this is an assault rifle, which would ideally deal much more damage (around 2.5 hearts, for scale) and would be automatic. Some very specific rifles also have a 3-burst mode, but (guess what?) that's the same as our SMG.

We already have the most common types of guns, except for a sniper rifle, and this PR serves to fill this gap. But the "plain ol'" rifle doesn't belong to any of these categories, and instead, falls under a very niche category of single-fire semi-automatic rifles. These WWI-era rifles resemble bolt-action sniper rifles at first glance, but are single-fire and have no scope. One such popular rifle is the [Lee-Enfield rifle](https://en.wikipedia.org/wiki/Lee–Enfield), linked here for the purpose of example. To summarise, these rifles aren't fun to play with at all. :)

MinetestSam commented:
I would suggest making the better sniper rifle only craftable. This would emphasise importance of crafting and would benefit those who craft stuff. This would also encourage players to craft which is what we should be looking for.

fjmng commented:
Tested and works

ClobberXD commented:
Thanks for testing!

> I would suggest making the better sniper rifle only craftable. This would emphasise importance of crafting and would benefit those who craft stuff.

Agreed. Will do.

LoneWolfHT commented on mods/pvp/sniper_rifles/init.lua:
```diff
@@ -0,0 +1,111 @@
+
+minetest.register_globalstep(function(dtime)
+	timer = timer + dtime
+	if timer < 0.05 then
```
Default server step is 0.1. This code seems pointless?

ClobberXD replied:
> **Default** server step is 0.1

If the server step duration is higher than this timer, this function will run every server step. If the server step has been decreased to less than 0.05, then this will run only once for multiple steps. That being said, 0.05 is too less, and a timer of 0.1s is more reasonable and less prone to lag.

ClobberXD commented:
This PR will be updated to use the brand-spanking-new `ObjectRef:set_fov` API method instead of the `zoom_fov` player property. This would allow for much cooler features like more robust zoom, scope-toggling instead of having to hold down the zoom key, etc.

LoneWolfHT commented:
Oh, just remembered another thing: It might look better if the guns were 16x16 too. I could have a go at that if you want

ClobberXD commented:
Feel free to have a go at it. Good luck. :)

LoneWolfHT commented:
16x16 was too little. So I set up some code to double the wield size of the sniper rifle and went with 32x32, it looks a lot better now IMO. I think we need to do this with the other weapons too at some point.
PR to this PR coming soon

LoneWolfHT commented:
https://github.com/ClobberXD/capturetheflag/pull/5

LoneWolfHT commented:
Ended up doing 32x32 but I think it's still slightly better, I want to go and redo the MTG textures now lol

ClobberXD commented:
Merged ClobberXD#5

ClobberXD commented:
## Update

- Mod uses `ObjectRef:set_fov` instead of the `zoom_fov` object property.
- The code has been compactified a great deal due to the aforementioned change.
- The actual rifle registrations have been moved to a separate file for refactoring purposes.
- A random message has been added to inform players about the ability to toggle scope using RMB/double-tap when wielding a sniper rifle.
- The 7.62mm rifle has a 4x scope, while the magnum rifle has an 8x scope. FOV multipliers are open to changes.

****

Coming soon: Scope overlay textures!

ClobberXD commented:
It's pretty hard to ensure that the scope overlay texture fits all screen sizes and more importantly, aspect ratios. To make sure this fits on all standard PC screens, I used a 1920x1080 image, positioned it in the centre of the screen, and scaled it by 100% in both axes, for my now-outdated [sniper](https://github.com/ClobberXD/sniper/blob/1126a6ebf2f80ce5798d02a4058b833e62c0aba7/api.lua#L38) mod. That'd allow the scope overlay to fit on all **16:9** screens.

However, the problem arises when aspect ratio starts changing, which is the case with certain laptops and of course, Android devices. To try and fix this, I made the overlay image 1920x1920 (1:1) and tried specifying the scale in only the x-axis to start out with, but the HUD element doesn't display at all. Even if that did work, we'd have to figure out how to detect the smaller dimension to fit the image to. All in all, static images aren't going to work. I'm planning to give dynamic (a.k.a on-the-fly) HUD elements a shot, but I'm not too sure about that either. It'd be great if we could have a more low-level HUD element that can be used to manually *draw* HUD elements on the screen.

ClobberXD commented:
## Updates

- Player is automatically "scoped-out" when their wielded item changes.
- Sniper rifles can now (only!) be used by sharp shooters.
- No scope overlay textures at the moment, sorry!

### TODO

- Make Magnum rifle craftable-only, or at least very rare in treasure chests. Recipe suggestions welcome.

ClobberXD commented:
## Updates

- Tweaked treasure spawning rates. The Magnum rifle is now ultra-rare to find in treasure chests.
- Rifles spawning as treasures will be loaded, instead of being empty.
- Both rifles can now be crafted. Recipes:
  - 7.62mm rifle:
    - Steel block x1
    - Bronze ingots x5
    - Mese crystal x1
    - Wood planks x1
  - Magnum rifle
    - Steel block x1
    - Bronze block x1
    - Diamond x1
    - Wood planks x1

****

With that out of the way, this PR is ready for review. :tada:

rubenwardy commented:
Looks good to me!

Would you be able to summarise how far you got with adding a HUD element for the scope?
